### PR TITLE
Add exposure events and status

### LINKED ISSUE #217: Add exposure events and status to experiment details view
In the experiments details view, we forgot to include the `exposure_events` and `status`! :) 

Let's add `exposure_events` to the Audience card with the current grouping design. Set as low priority for now in case we change the grouping design.

And let's add `status` as it's own row to the General card right above `Dates`. We can use the same `ExperimentStatus` component that is used in the Experiments List.

### PR DESCRIPTION
**Pull request by withinboredom**
resolves #217 

This adds exposure events and the status to the experiment details pane. I chose to use several layers of `LabelValueTable` because this sort-of provides a "tree-like" structure which is the best way to visualize the data IMHO. I'm happy to use a different component and/or a different approach. I also chose to keep the `ExposureEventsTable` as a non-reusable component because 1) it can be easily refactored into one if we chose to and 2) it's incredibly simple such that I'm not sure we need to make it it's own file yet. 3) I'm not sure what the conventions are around here yet :)

## How has this been tested?

- Automated tests that cover added/changed functionality
- Manual testing with production data (locally)
- Manual testing with mock data (locally or on Vercel)
- Additional manual testing instructions (please specify):

### CODE REVIEW

**Comment by jessie-ross** [comment]
The code and code decisions looks good, but I'm not too sure about the design.

Screen Shot 2020-09-01 at 11 53 06 am

I can some potential issues here:
- Event names can be quite long, that can really mess with everything.
- There will be a lot of custom text - every instance of 'string', 'additionalPropX' could be any text. This means there can be a big blob of unstructured text over that space - it is hard to tell what you are looking at, are `additionalProps` the events?
- Bold text so far hasn't meant user-data but rather serves as a header for user-data, this could be confusing.

(I do still think JSON is a good representation for these especially being so data-intense and having custom object properties, but I don't think I could convince @yanirs.)

This is quite a hard component to style as we are presenting a lot of complex and structured information within a larger also complex display.

I'm not too sure these will work but how about:
- Not using LabelValueTable, it doesn't do much anyway, very easy to inline and will make the below simpler.
- Having `eventName` on its own row above the properties. Have `<strong>Event:</strong> event_name`
- Use monospace fonts, we should be using it for all user input: `fontFamily: theme.custom.fonts.monospace`
- Having props below it but indented(? not crazy about this but might be good for now).
- It also looks like you've run into a problem of deeply nested tables not showing the borders between items, might be worth fixing, but I could fix in a future PR if it is too troublesome.

**Comment by yanirs** [comment]
> (I do still think JSON is a good representation for these especially being so data-intense and having custom object properties, but I don't think I could convince @yanirs.)

Maybe JSON would work for the props? We can also consider moving exposure events to a separate panel, since they're technically a part of the metric calculation rather than of the audience.

Another option is to mimic the way we have it in funnels (props under the event name and greyed). Example:

image
 
Anyway, I think we can push it with minor tweaks and iterate on the design. @aaronyan may also have some design ideas.

**Comment by jessie-ross** [comment]
Yeah I think greying the props might help, we might not need the 'Event:' text if that's the case.

I also like the simple `key: value` style so the props themselves don't have to be in a table.

**Comment by withinboredom** [comment]
Maybe something like this:

image

PS. Last commit will fail tests (I had a hard stop so will fix here shortly)

**Comment by jessie-ross** [comment]
Yeah that is what I mean as far as the layout.

I am not sure if you mean that image literally but if it is I would say let's reduce it a bit further:
- No need for 'Event:' now.
- The underline is new design language, it might be a bit confusing.
- Let's stick to greyscale to keep it simple and not distracting from other parts of the page.
- With the grey props and use of colons (`:`), it should be obvious the event is an event, I don't think it needs the indent any more.

**Comment by jessie-ross** [comment]
This is what it might look with the points above:

Screen Shot 2020-09-01 at 11 59 18 pm

It will look best once monospace is uniform throughout.

**Comment by jessie-ross** [comment]
Ahh, it is also a bit strange that Exposure Events is with Audience, but why not for now. 
I would still love to have a Meta-Parameters section.

File: components/AudiencePanel.tsx [hunk 1]
```diff
@@ -52,6 +53,45 @@ const useStyles = makeStyles(() =>
...
+}))
+
+function ExposureEventsTable({ experiment }: { experiment: ExperimentFull }) {
+  const classes = eventStyles();
+
+  const renderProps = (props: {} | null | undefined) => {
+    if(props === undefined || props === null || props == {}) return null
+    const asArray = Object.entries(props)
+    return <ul className={classes.propsList}>
+      {asArray.map(entry => <li className={classes.entry}>{entry[0]}:{entry[1]}</li>)}
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:79]: It is good to use list destructuring for this sort of thing:
```
      {asArray.map(([key, val]) => <li className={classes.entry}>{key}:{val}</li>)}

File: components/AudiencePanel.tsx [hunk 2]
```diff
@@ -52,6 +53,45 @@ const useStyles = makeStyles(() =>
...
+    </ul>
+  }
+
+  return (
+    <>
+      {experiment.exposureEvents?.map((ev) =>
+              <Typography key={ev.event}>
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:86]: Might want to add a style here for multiple events, ensuring a border between and a nice padding on both sides (`theme.spacing(2 or 3 or 4)` is your friend here.)

File: components/AudiencePanel.tsx [hunk 3]
```diff
@@ -52,6 +53,45 @@ const useStyles = makeStyles(() =>
...
+    marginTop: 0
+  }
+}))
+
+function ExposureEventsTable({ experiment }: { experiment: ExperimentFull }) {
+  const classes = eventStyles();
+
+  const renderProps = (props: {} | null | undefined) => {
+    if(props === undefined || props === null || props == {}) return null
+    const asArray = Object.entries(props)
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:77]: `eventPropertyEntries` might be a nice name here

File: components/AudiencePanel.tsx [hunk 4]
```diff
@@ -52,6 +53,45 @@ const useStyles = makeStyles(() =>
...
+  propsList: {
+    marginTop: 0
+  }
+}))
+
+function ExposureEventsTable({ experiment }: { experiment: ExperimentFull }) {
+  const classes = eventStyles();
+
+  const renderProps = (props: {} | null | undefined) => {
+    if(props === undefined || props === null || props == {}) return null
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:76]: No need to check `props ==(=?) {}` as the rest of the code will just work.

This means this can be collapsed quite a bit:

`{exposureEvents && <ul> Object.entries(exposureEvents).map(([key, val]) => <li> key:val </li> </ul>}`

File: components/AudiencePanel.tsx [hunk 5]
```diff
@@ -52,6 +53,45 @@ const useStyles = makeStyles(() =>
...
+  },
+  propsList: {
+    marginTop: 0
+  }
+}))
+
+function ExposureEventsTable({ experiment }: { experiment: ExperimentFull }) {
+  const classes = eventStyles();
+
+  const renderProps = (props: {} | null | undefined) => {
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:75]: A good general rule is that any function that produces JSX should be a component.

This function signature is particularly strange as `props` doesn't follow the usual conventions, `props` is always an object and it normally contains data on a property of `props`.

In this case I would just inline it.

File: components/AudiencePanel.tsx [hunk 6]
```diff
@@ -52,6 +53,45 @@ const useStyles = makeStyles(() =>
...
+    if(props === undefined || props === null || props == {}) return null
+    const asArray = Object.entries(props)
+    return <ul className={classes.propsList}>
+      {asArray.map(entry => <li className={classes.entry}>{entry[0]}:{entry[1]}</li>)}
+    </ul>
+  }
+
+  return (
+    <>
+      {experiment.exposureEvents?.map((ev) =>
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:85]: Might also want to show "No exposure events" if there are no exposure events.

**Comment by withinboredom** [comment]
Thanks for all the amazing feedback @jessie-ross!

image

Much, much better. Still probably not perfect, but it looks great to my untrained eye... ha!

File: components/AudiencePanel.tsx [hunk 7]
```diff
@@ -52,6 +53,58 @@ const useStyles = makeStyles(() =>
...
+    },
+    eventList: {
+      '& p:not(:first-child)': {
+        paddingTop: theme.spacing(2),
+      },
+      '& p': {
+        paddingBottom: theme.spacing(2),
+      },
+      '& p:not(:last-child)': {
+        borderBottom: '1px solid rgb(224,224,224)',
```
Comment (withinboredom) [inline components/AudiencePanel.tsx:78]: I'm not sure if this is the proper way to do colors here :thinking:

**Comment by withinboredom** [comment]
I have no idea what the tests are complaining about, it looks like everything passed to me.

**Comment by yanirs** [comment]
@withinboredom it looks much better!

> Ahh, it is also a bit strange that Exposure Events is with Audience, but why not for now.
> I would still love to have a Meta-Parameters section.

@jessie-ross yeah, it being in Audience isn't great, but good enough for now. Shuffling things around might be a good thing to tackle in the next milestone &ndash; can note it in our planning meeting. :slightly_smiling_face: 

> I have no idea what the tests are complaining about, it looks like everything passed to me.

@withinboredom it's the coverage:
```
Jest: "global" coverage threshold for statements (100%) not met: 99.75%
Jest: "global" coverage threshold for branches (100%) not met: 98.33%
Jest: "global" coverage threshold for lines (100%) not met: 99.74%
Jest: "global" coverage threshold for functions (100%) not met: 99.07%
```

Adding a test case for `AudiencePanel` with non-null exposure events should fix it.

File: components/AudiencePanel.tsx (hunk 7 above)
Comment (jessie-ross) [reply components/AudiencePanel.tsx:78]: Generally prefer to use the "#" notation but it doesn't really matter all that much 😉

File: components/AudiencePanel.tsx [hunk 8]
```diff
@@ -52,6 +53,58 @@ const useStyles = makeStyles(() =>
...
+              {' '}
+              {Object.entries(ev.props).map(([key, val]) => (
+                <li key={key + val} className={classes.entry}>
+                  {key}: {val}
+                </li>
+              ))}
+            </ul>
+          )}
+        </Typography>
+      )) ?? <Typography>No exposure events defined</Typography>}
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:103]: There are two cases that could give no exposure events:
- `exposureEvents === null`
- `exposureEvents === [] `

It looks to me like only the first is covered here:

Screen Shot 2020-09-02 at 11 27 15 am

Comment (withinboredom) [reply components/AudiencePanel.tsx:103]: I'll fix it when I add tests (I had assumed we'd already had tests for all the cases).

File: components/AudiencePanel.tsx [hunk 9]
```diff
@@ -91,7 +91,7 @@ function ExposureEventsTable({ experiment }: { experiment: ExperimentFull }) {
           <span className={classes.eventName}>{ev.event}</span>
           {ev.props &&
             Object.entries(ev.props).map(([key, val]) => (
-              <span key={key + val} className={classes.entry}>
+              <span key={key} className={classes.entry}>
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:94]: Nice! I don't think it would have mattered either way but this is what I would call the preferred choice.

Comment (withinboredom) [reply components/AudiencePanel.tsx:94]: I'm more concerned with more than one event in the array having the same name and then the same key. Dunno what to do about that...

Comment (jessie-ross) [reply components/AudiencePanel.tsx:94]: Yep that's a good point, we could add it to our yup schema very easily later on, but is it much of a problem if that happens? Do we track exposures separately or are they combined?

Comment (withinboredom) [reply components/AudiencePanel.tsx:94]: In theory, there won't be duplicate event names (I would guess that behavior is undefined in the analysis). It's probably a non-issue.

File: components/AudiencePanel.tsx [hunk 10]
```diff
@@ -52,6 +53,54 @@ const useStyles = makeStyles(() =>
...
+      '& p:not(:last-child)': {
+        borderBottom: '1px solid rgb(224,224,224)',
+      },
+    },
+  }),
+)
+
+function ExposureEventsTable({ experiment }: { experiment: ExperimentFull }) {
+  const classes = eventStyles()
+  const exposureEvents =
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:82]: Might be clearer for a `const isExposureEvents = experiment.exposureEvents && experiment.exposureEvents !== []`

**Review by jessie-ross — APPROVED** [review]
See my last comment but otherwise good to go! 🚀

File: components/AudiencePanel.tsx (hunk 10 above)
Comment (withinboredom) [reply components/AudiencePanel.tsx:82]: I thought about that, but for whatever reason decided to be extra defensive. I'll update it to be more permissive for clarity.

Comment (withinboredom) [reply components/AudiencePanel.tsx:82]: Ah, that doesn't work :) 

```
[] && [] !== []
true && false
true
```

Which isn't what we want.

Comment (withinboredom) [reply components/AudiencePanel.tsx]: That should be much better now :)

File: components/AudiencePanel.tsx [hunk 11]
```diff
@@ -77,26 +77,26 @@ const eventStyles = makeStyles(() =>
...
-          <span className={classes.eventName}>{ev.event}</span>
-          {ev.props &&
-            Object.entries(ev.props).map(([key, val]) => (
-              <span key={key} className={classes.entry}>
-                {key}: {val}
-              </span>
-            ))}
-        </Typography>
-      )) ?? <Typography>No exposure events defined</Typography>}
+      {exposureEvents && exposureEvents.length ? (
```
Comment (jessie-ross) [inline components/AudiencePanel.tsx:85]: Nice! Yeah the ternary is the right way to go here.
